polygon/sync/difficulty.go

```diff
@@ -0,0 +1,54 @@
+package sync
+
+import (
+	lru "github.com/hashicorp/golang-lru/arc/v2"
+	"github.com/ledgerwatch/log/v3"
+
+	"github.com/ledgerwatch/erigon-lib/chain"
+	libcommon "github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon/consensus/bor"
+	heimdallspan "github.com/ledgerwatch/erigon/consensus/bor/heimdall/span"
+	"github.com/ledgerwatch/erigon/consensus/bor/valset"
+	"github.com/ledgerwatch/erigon/core/types"
+)
+
+type DifficultyCalculator interface {
+	HeaderDifficulty(header *types.Header) (uint64, error)
+}
+
+type difficultyCalculatorImpl struct {
+	borConfig       *chain.BorConfig
+	span            *heimdallspan.HeimdallSpan
+	signaturesCache *lru.ARCCache[libcommon.Hash, libcommon.Address]
+}
+
+func NewDifficultyCalculator(
+	borConfig *chain.BorConfig,
+	span *heimdallspan.HeimdallSpan,
+) DifficultyCalculator {
+	signaturesCache, err := lru.NewARC[libcommon.Hash, libcommon.Address](4096)
+	if err != nil {
+		panic(err)
+	}
+	return &difficultyCalculatorImpl{
+		borConfig:       borConfig,
+		span:            span,
+		signaturesCache: signaturesCache,
+	}
+}
+
+func (impl *difficultyCalculatorImpl) HeaderDifficulty(header *types.Header) (uint64, error) {
+	signer, err := bor.Ecrecover(header, impl.signaturesCache, impl.borConfig)
```

Comment:
what if the signer is not in the span selected producers/validator set?

Reply:
in this case the Difficulty() call below will return an error:
`errors.New("ValidatorSet.Difficulty: signer index not found")`
(and this error bubbles up from Connect())
I've seen that there's an explicit check in other places, but I felt it is redundant